routers/api/v1/user/gpg_key.go

```diff
@@ -132,6 +133,11 @@ func GetGPGKey(ctx *context.APIContext) {
 
 // CreateUserGPGKey creates new GPG key to given user by ID.
 func CreateUserGPGKey(ctx *context.APIContext, form api.CreateGPGKeyOption, uid int64) {
+	if setting.Admin.UserDisabledFeatures.Contains(setting.UserFeatureManageGPGKeys) {
+		ctx.NotFound("Not Found", fmt.Errorf("gpg keys setting is not allowed to be visited"))
+		return
+	}
```

Comment:
Should we move this check into the middleware to keep the logic in the router on topic?

Reply:
I did that on #20549 but @wxiaoguang think it's better to move them into the router function body.

Reply:
Do you mean this comment? https://github.com/go-gitea/gitea/pull/29275#discussion_r1496895057

At that time:

1. Only one handler needs such check
2. It bloats the "web.go", do we really want to fill more and more function into "web.go"?

For this time: it is still the only way: put checks into `routers/web/user/setting/keys.go`, it is impossible to "move into middleware" for it.


My reviews never means "it must do so", I am always open for better ideas.


So, for this problem: what is the consistent and maintainable approach in your mind?

Reply:
Something like this for the middleware:
```go
func featureEnabled(feature /*...*/string) func(ctx *context.Context) {
	if setting.Admin.UserDisabledFeatures.Contains(feature) {
		ctx.Error(http.StatusNotFound)
		return
	}
}
```
But that does not work well with the method with the `type` parameter. Repeating the same code multiple times is not really good either.

Reply:
And by the way, I think my concern for "testing" is still not addressed there https://github.com/go-gitea/gitea/pull/29275#discussion_r1496912355

Reply:
Not really no way but we need to use one unpleasant way.